### PR TITLE
Update Fn: Executor New Deployment

### PR DESCRIPTION
**Pull request by vishal-biyani**
This is PR for #465. 
Yet to test extensively, will remove "in progress" flag once ready for review.

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/504)

### CODE REVIEW

File: fission/function.go [hunk 1]
```diff
@@ -373,6 +379,37 @@ func fnUpdate(c *cli.Context) error {
 	entrypoint := c.String("entrypoint")
 	buildcmd := c.String("buildcmd")
 	force := c.Bool("force")
+	secretName := c.String("secret")
```
Comment (soamvasani) [inline fission/function.go]: Did some mis-merge happen here? Why is this change in this PR?

**Comment by soamvasani** [comment]
I'm a bit confused by the size of this change. Are there some other changes merged in here?  Can you add a detailed description?

File: fission/function.go (hunk 1 above)
Comment (vishal-biyani) [reply fission/function.go]: So I could not find the code for updating secrets and configmaps, so added this.

**Comment by vishal-biyani** [comment]
I will add a detailed description and the testing details

**Comment by vishal-biyani** [comment]
Since this change looks a bit more than what it should be, here are some details. All line numbers are on the right-hand side of change:

- newdeploy.go L115 - L238 : This is actually existing code being pulled into a separate method. The intention is to separate spec construction from creation of deployment, so updates can be handled easily. Not sure why Git can not understand this better.

- newdeploy.go L240 - L271 : This method takes care of overriding the resources specified in function over the one adopted from the environment.

- newdeploy.go L320 - L330: A read operation for HPA and similar CRUD operations added nearby

- newdeploymgr.go L311 - L444: This is the `meat` of this change really, what the PR was meant for! Handles updates of function for various use cases. There is some room for improvement in this long method.

- environment.go L208 - L257 : Most of the changes are small, additional checks on user input

- function.go L382 - L412 : Strangely, the config maps and secrets updates were not handled earlier. I tried to find it but could not, so ended up adding in this PR! 

@soamvasani @life1347 Please use above as a reference when you review.

File: fission/function.go (hunk 1 above)
Comment (life1347) [reply fission/function.go]: Let me move this change to another PR.

**Comment by life1347** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[executor/newdeploy/newdeploymgr.go, line 171 at r4](https://reviewable.io:443/reviews/fission/fission/504#-L6GIFsA4KiTlQ2m0hz0:-L6GIFsA4KiTlQ2m0hz1:b-pspar) ([raw file](https://github.com/fission/fission/blob/e27c6c3830de590c187fcc57f20348f9628ea86d/executor/newdeploy/newdeploymgr.go#L171)):*
> by controller
by informer controller

Though it's not a big problem. But in fact, I am more prefer to call fnUpdate here as well for code consistency.
We can add an additional field called `oldFn` in `fnRequest` to solve this problem.
```
fnRequest struct {
	reqType         requestType
	fn                   *crd.Function
	oldfn              *crd.Function
	responseChannel chan *fnResponse
}
```

---

*[executor/newdeploy/newdeploymgr.go, line 368 at r4](https://reviewable.io:443/reviews/fission/fission/504#-L6GOSC6ra7zOuu2r0IH:-L6GOSC6ra7zOuu2r0II:btu8kc7) ([raw file](https://github.com/fission/fission/blob/e27c6c3830de590c187fcc57f20348f9628ea86d/executor/newdeploy/newdeploymgr.go#L368)):*
> ```Go
> 		}
> 
> 		err = deploy.updateDeployment(deployment)
> ```

Can we move deploy.updateDeployment to the end of fnUpdate? 
Cause in worst cases, a function's deployment will be updated for 3 times. It increases the delay for the latest function to be accessed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/504)*

**Comment by vishal-biyani** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[executor/newdeploy/newdeploymgr.go, line 171 at r4](https://reviewable.io:443/reviews/fission/fission/504#-L6GIFsA4KiTlQ2m0hz0:-L6GQ8YBs5Sq57nRShC-:b-tvgxfj) ([raw file](https://github.com/fission/fission/blob/e27c6c3830de590c187fcc57f20348f9628ea86d/executor/newdeploy/newdeploymgr.go#L171)):*
<details><summary><i>Previously, life1347 (Ta-Ching Chen) wrote…</i></summary><blockquote>

> by controller
by informer controller

Though it's not a big problem. But in fact, I am more prefer to call fnUpdate here as well for code consistency.
We can add an additional field called `oldFn` in `fnRequest` to solve this problem.
```
fnRequest struct {
	reqType         requestType
	fn                   *crd.Function
	oldfn              *crd.Function
	responseChannel chan *fnResponse
}
```
</blockquote></details>

Yeah, I thought about it too, but the update will not be explicitly called. Update here is only triggered by a update in function - which is why I left this blank. Maybe add a comment and leave it in current state?

---

*[executor/newdeploy/newdeploymgr.go, line 368 at r4](https://reviewable.io:443/reviews/fission/fission/504#-L6GOSC6ra7zOuu2r0IH:-L6GUIEl52Zg5j0FljX_:b-uejjkq) ([raw file](https://github.com/fission/fission/blob/e27c6c3830de590c187fcc57f20348f9628ea86d/executor/newdeploy/newdeploymgr.go#L368)):*
<details><summary><i>Previously, life1347 (Ta-Ching Chen) wrote…</i></summary><blockquote>

Can we move deploy.updateDeployment to the end of fnUpdate? 
Cause in worst cases, a function's deployment will be updated for 3 times. It increases the delay for the latest function to be accessed.
</blockquote></details>

Great point ta-ching! This suggestion removes some code and simplifies the whole process a lot more

---

*[fission/function.go, line 382 at r1](https://reviewable.io:443/reviews/fission/fission/504#-L5s0OTCF-vdRXMN97os-r1-382:-L6GQ6-OmUIzlLoKj407:b-896fix) ([raw file](https://github.com/fission/fission/blob/44ff1ddafbb95a28dc478e3ca4c08f6684d1e878/fission/function.go#L382)):*
<details><summary><i>Previously, life1347 (Ta-Ching Chen) wrote…</i></summary><blockquote>

Let me move this change to another PR.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/504)*

**Comment by soamvasani** [comment]
Reviewed 1 of 6 files at r1, 1 of 2 files at r3, 2 of 2 files at r4, 1 of 1 files at r5.
Review status: 5 of 6 files reviewed at latest revision, 21 unresolved discussions.

---

*[executor/newdeploy/newdeploy.go, line 105 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IL2X7GPAhMGashpv_:-L6IL2XFL6v6_vPJ3U2o:b-bjumbe) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploy.go#L105)):*
> DeletePropagationForeground

Should this be foreground or background?

---

*[executor/newdeploy/newdeploy.go, line 240 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6ILFcZj9GVkyNVr8ly:-L6ILFcZj9GVkyNVr8lz:b-ysr28q) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploy.go#L240)):*
> ```Go
> }
> 
> func (deploy *NewDeploy) getResources(env *crd.Environment, fn *crd.Function) v1.ResourceRequirements {
> ```

please add a function comment

---

*[executor/newdeploy/newdeploymgr.go, line 263 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IMQPxZiVG1cLASC_6:-L6IMQQ-WR8VSKCRQxG4:bhtefdo) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L263)):*
> ```Go
> 	if err != nil {
> 		log.Printf("error creating the HPA %v: %v", objName, err)
> 		return fsvc, errors.Wrap(err, "error creating HorizontalPodAutoscaler")
> ```

would be helpful to add same message that you logged

---

*[executor/newdeploy/newdeploymgr.go, line 327 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IMbqRshzQYeeVE7pN:-L6IMbqRshzQYeeVE7pO:b-4qc84g) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L327)):*
> ```Go
> 	if oldFn.Spec.InvokeStrategy != newFn.Spec.InvokeStrategy {
> 
> 		if newFn.Spec.InvokeStrategy.ExecutionStrategy.ExecutorType == fission.ExecutorTypePoolmgr &&
> ```

shouldn't this be ... != ExecutorTypeNewdeploy

---

*[executor/newdeploy/newdeploymgr.go, line 335 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IN6EWwalNB7aiw5am:-L6IN6EWwalNB7aiw5an:bnj7hl0) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L335)):*
> ```Go
> 		}
> 
> 		if oldFn.Spec.InvokeStrategy.ExecutionStrategy.ExecutorType == fission.ExecutorTypePoolmgr &&
> ```

same as above

---

*[executor/newdeploy/newdeploymgr.go, line 347 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6INCa6mbznx3frrSZ2:-L6INCa6mbznx3frrSZ3:bljp041) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L347)):*
> ```Go
> 		hpa, err := deploy.getHpa(newFn)
> 		if err != nil {
> 			log.Printf("error getting HPA while updating function: %v", err)
> ```

was it intentional to ignore the error?

---

*[executor/newdeploy/newdeploymgr.go, line 376 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6INbdGj_QLY5QdAEpw:-L6INbdGj_QLY5QdAEpx:bule8g0) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L376)):*
> ```Go
> 
> 	if oldFn.Spec.Package.PackageRef != newFn.Spec.Package.PackageRef ||
> 		oldFn.Spec.Environment != newFn.Spec.Environment {
> ```

why is the environmentref compared twice?

---

*[executor/newdeploy/newdeploymgr.go, line 406 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6INxKnZ4Ee8YBRqyIG:-L6INxKnZ4Ee8YBRqyIH:bgz5u4j) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L406)):*
> ```Go
> 			Get(newFn.Spec.Environment.Name)
> 		if err != nil {
> 			log.Printf("failed to get environment while updating function: %v", err)
> ```

ignoring error?

---

*[executor/newdeploy/newdeploymgr.go, line 413 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6INyuAVPcEFlauGPn1:-L6INyuAVPcEFlauGPn2:bgz5u4j) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L413)):*
> ```Go
> 		newDeployment, err := deploy.getDeploymentSpec(newFn, env, deployName, deployLabels)
> 		if err != nil {
> 			log.Printf("failed to get new deployment spec while updating function: %v", err)
> ```

ignoring error?

---

*[executor/newdeploy/newdeploymgr.go, line 417 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IO-M2eeHZBmMYgfjF:-L6IO-M4Ag_U6QqhHdrb:bgz5u4j) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L417)):*
> ```Go
> 		err = deploy.updateDeployment(newDeployment)
> 		if err != nil {
> 			log.Printf("failed to update deployment while updating function: %v", err)
> ```

ignoring error?

---

*[fission/function.go, line 211 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IP3sDf4SMxUQhJMCg:-L6IP3sEoTS5pCd62tqz:b-uu35r7) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L211)):*
> ```Go
> 	//TODO Warn user about resources at fn level overriding the env resources
> 	if c.IsSet("mincpu") || c.IsSet("maxcpu") || c.IsSet("minmemory") || c.IsSet("maxmemory") {
> 		warn("Resources specified for function will override the once specified for environment")
> ```

I don't think we need this warning...?  Function overriding environment's resources is user's expected behaviour, I think.  It would be useful to have the opposite warning, when function resource requirements are specified but cannot be met (because we're using poolmgr).

---

*[fission/function.go, line 404 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IP_bnTYrnvVjpoPat:-L6IP_bnTYrnvVjpoPau:b-dsrz7f) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L404)):*
> ```Go
> 	}
> 
> 	pkg, err = client.PackageGet(&metav1.ObjectMeta{
> ```

double packageGet?

---

*[fission/function.go, line 444 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IPca8YWGVjfTJfE-3:-L6IPca8YWGVjfTJfE-4:bq66zrf) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L444)):*
> ```Go
> 
> 	if c.IsSet("mincpu") || c.IsSet("maxcpu") || c.IsSet("minmemory") || c.IsSet("maxmemory") {
> 		warn("Resources specified for function will override the once specified for environment")
> ```

the ones*

but also this warning is not really needed, I think

---

*[fission/function.go, line 450 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IPktGrZckc5R3I4eI:-L6IPktGrZckc5R3I4eJ:b-h8yf4p) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L450)):*
> ```Go
> 	function.Spec.InvokeStrategy.ExecutionStrategy.TargetCPUPercent = getTargetCPU(c)
> 
> 	if c.IsSet("minscale") {
> ```

can you file an issue so we can move this into CRD validation at some point?

---

*[fission/function.go, line 453 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IQquBvQk6c6igLJoY:-L6IQquBvQk6c6igLJoZ:bf1l1f) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L453)):*
> ```Go
> 		minscale := c.Int("minscale")
> 		if c.IsSet("maxscale") && minscale > c.Int("maxscale") {
> 			fatal("Minscale can not be greater than maxscale")
> ```

helpful to say what the actual number is for min/max scale

---

*[fission/function.go, line 457 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IQezP72tWfZ9Fk6A3:-L6IQezQrSsNNz58Pd4_:b-rxgqwh) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L457)):*
> ```Go
> 		if function.Spec.InvokeStrategy.ExecutionStrategy.ExecutorType != fission.ExecutorTypePoolmgr &&
> 			minscale > function.Spec.InvokeStrategy.ExecutionStrategy.MaxScale {
> 			fatal("Minscale can not be greater than maxscale")
> ```

more helpful error message: "Min scale cannot be greater than existing function's max scale (which is X)"

---

*[fission/function.go, line 465 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IQx36OW-BSyuPj4n7:-L6IQx36OW-BSyuPj4n8:b7xdzkm) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L465)):*
> ```Go
> 		maxscale := c.Int("maxscale")
> 		if maxscale < function.Spec.InvokeStrategy.ExecutionStrategy.MinScale {
> 			fatal("Function's minscale can not be greater than --maxscale")
> ```

be consistent with using -- or not

---

*[test/tests/test_backend_nd_update.sh, line 66 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IRCirbHzW-w6Ivlg4:-L6IRCirbHzW-w6Ivlg5:b7rs2gm) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/test/tests/test_backend_nd_update.sh#L66)):*
> ```Shell
> }
> 
> # This test only tests one path of execution which can be easily validated from outside cluster
> ```

which path? please specify :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/504)*

**Comment by soamvasani** [comment]
Reviewed 1 of 6 files at r1.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/504)*

**Comment by vishal-biyani** [comment]
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[executor/newdeploy/newdeploy.go, line 105 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IL2X7GPAhMGashpv_:-L6LtfCoeo49bujXUPc1:b-7egydw) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploy.go#L105)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

> DeletePropagationForeground

Should this be foreground or background?
</blockquote></details>

This is not exactly a deletion in foreground - so I am going to keep it as it is for now.

---

*[executor/newdeploy/newdeploy.go, line 240 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6ILFcZj9GVkyNVr8ly:-L6LuWWhiG7uK9zxEdVC:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploy.go#L240)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

please add a function comment
</blockquote></details>

Done.

---

*[executor/newdeploy/newdeploymgr.go, line 171 at r4](https://reviewable.io:443/reviews/fission/fission/504#-L6GIFsA4KiTlQ2m0hz0:-L6LuXaL87fA62fWujQJ:b-896fix) ([raw file](https://github.com/fission/fission/blob/e27c6c3830de590c187fcc57f20348f9628ea86d/executor/newdeploy/newdeploymgr.go#L171)):*
<details><summary><i>Previously, vishal-biyani (Vishal) wrote…</i></summary><blockquote>

Yeah, I thought about it too, but the update will not be explicitly called. Update here is only triggered by a update in function - which is why I left this blank. Maybe add a comment and leave it in current state?
</blockquote></details>

Done.

---

*[executor/newdeploy/newdeploymgr.go, line 263 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IMQPxZiVG1cLASC_6:-L6Ly0MH2SjSglygrzM8:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L263)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

would be helpful to add same message that you logged
</blockquote></details>

Done.

---

*[executor/newdeploy/newdeploymgr.go, line 327 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IMbqRshzQYeeVE7pN:-L6LyMluqv5NsBJejt8p:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L327)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

shouldn't this be ... != ExecutorTypeNewdeploy
</blockquote></details>

Done.

---

*[executor/newdeploy/newdeploymgr.go, line 335 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IN6EWwalNB7aiw5am:-L6LzOIND6rOM81ZEjku:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L335)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.

---

*[executor/newdeploy/newdeploymgr.go, line 347 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6INCa6mbznx3frrSZ2:-L6LztwSew12pT_vVwGd:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L347)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

was it intentional to ignore the error?
</blockquote></details>

Done.

---

*[executor/newdeploy/newdeploymgr.go, line 376 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6INbdGj_QLY5QdAEpw:-L6M03E4yuE82xW_cTxr:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L376)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

why is the environmentref compared twice?
</blockquote></details>

Done.

---

*[executor/newdeploy/newdeploymgr.go, line 406 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6INxKnZ4Ee8YBRqyIG:-L6M04HO8UKolHmEtAEt:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L406)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

ignoring error?
</blockquote></details>

Done.

---

*[executor/newdeploy/newdeploymgr.go, line 413 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6INyuAVPcEFlauGPn1:-L6M068daQRnfm82552-:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L413)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

ignoring error?
</blockquote></details>

Done.

---

*[executor/newdeploy/newdeploymgr.go, line 417 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IO-M2eeHZBmMYgfjF:-L6M06VWQ9jRY-7qTYBz:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploymgr.go#L417)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

ignoring error?
</blockquote></details>

Done.

---

*[fission/function.go, line 211 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IP3sDf4SMxUQhJMCg:-L6M0LP1B1g60qwDwcLz:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L211)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

I don't think we need this warning...?  Function overriding environment's resources is user's expected behaviour, I think.  It would be useful to have the opposite warning, when function resource requirements are specified but cannot be met (because we're using poolmgr).
</blockquote></details>

Done.

---

*[fission/function.go, line 404 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IP_bnTYrnvVjpoPat:-L6M0qj-9Qg4Wal2IbIz:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L404)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

double packageGet?
</blockquote></details>

Done.

---

*[fission/function.go, line 444 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IPca8YWGVjfTJfE-3:-L6M0s1SqK0bVzwxi8Vl:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L444)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

the ones*

but also this warning is not really needed, I think
</blockquote></details>

Done.

---

*[fission/function.go, line 450 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IPktGrZckc5R3I4eI:-L6M1WyLflbafBNx1zi9:bffqjcl) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L450)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

can you file an issue so we can move this into CRD validation at some point?
</blockquote></details>

#481 Already does it.

---

*[fission/function.go, line 453 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IQquBvQk6c6igLJoY:-L6M3M-5-z15AisMFmSf:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L453)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

helpful to say what the actual number is for min/max scale
</blockquote></details>

Done.

---

*[fission/function.go, line 457 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IQezP72tWfZ9Fk6A3:-L6M3N3udDm8DIoNC95_:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L457)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

more helpful error message: "Min scale cannot be greater than existing function's max scale (which is X)"
</blockquote></details>

Done.

---

*[fission/function.go, line 465 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IQx36OW-BSyuPj4n7:-L6M3gMkcxnbtx2rXl0q:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L465)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

be consistent with using -- or not
</blockquote></details>

Done.

---

*[test/tests/test_backend_nd_update.sh, line 66 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IRCirbHzW-w6Ivlg4:-L6M4-hbDRB7rdCuU5ZO:b-896fix) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/test/tests/test_backend_nd_update.sh#L66)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

which path? please specify :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/504)*

**Comment by soamvasani** [comment]
LGTM after a few minor points--

---

Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[executor/newdeploy/newdeploy.go, line 105 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IL2X7GPAhMGashpv_:-L6NttMYXushPIm6euHY:b-yafl3m) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/executor/newdeploy/newdeploy.go#L105)):*
<details><summary><i>Previously, vishal-biyani (Vishal) wrote…</i></summary><blockquote>

This is not exactly a deletion in foreground - so I am going to keep it as it is for now.
</blockquote></details>

Please add a comment justifying the choice

---

*[executor/newdeploy/newdeploymgr.go, line 487 at r6](https://reviewable.io:443/reviews/fission/fission/504#-L6NuG8NJv7tB8Yld8rj:-L6NuG8NJv7tB8Yld8rk:b-dv9zdp) ([raw file](https://github.com/fission/fission/blob/8181a9e766889ad612bcf14980e5e7dbb66fb93c/executor/newdeploy/newdeploymgr.go#L487)):*
> ```Go
> }
> 
> func updateStatus(message string) {
> ```

This function's purpose if I understand it right is as a placeholder for updating the Function.Status in the future. That means it needs (a) to know whether there's an error (b) the reference to the function CRD. So, something like `func updateFunctionStatus(fn *crd.Function, err error)`

---

*[fission/function.go, line 211 at r5](https://reviewable.io:443/reviews/fission/fission/504#-L6IP3sDf4SMxUQhJMCg:-L6Nv_Rjbwuv5TPTRR5y:b-6axd9j) ([raw file](https://github.com/fission/fission/blob/70cc6708d17ebb7dbc31896403ee0b594eb2e432/fission/function.go#L211)):*
<details><summary><i>Previously, vishal-biyani (Vishal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Did you add a warning for when a function's resource requirements are specified but cannot be met? (Because of poolmgr being used as backend)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/504#-:-L6NvrrYgd_-3odBQnhM:bxh14hs)*

**Pull request merged**